flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/CloseableSubtaskGateway.java

```diff
@@ -0,0 +1,182 @@
+     * allow closing the gateway for this specific checkpoint.
+     *
+     * <p>This is the gateway's mechanism to detect situations where multiple coordinator
+     * checkpoints would be attempted overlapping, which is currently not supported (the gateway
+     * doesn't keep a list of events blocked per checkpoint). It also helps to identify situations
+     * where the checkpoint was aborted even before the gateway was closed (by finding out that the
+     * {@code currentCheckpointId} was already reset to {@code NO_CHECKPOINT}.
+     */
+    void markForCheckpoint(long checkpointId) {
+        checkRunsInMainThread();
+
+        if (currentCheckpointId != NO_CHECKPOINT && currentCheckpointId != checkpointId) {
+            throw new IllegalStateException(
+                    String.format(
+                            "Cannot mark for checkpoint %d, already marked for checkpoint %d",
+                            checkpointId, currentCheckpointId));
+        }
+        if (checkpointId > lastCheckpointId) {
+            currentCheckpointId = checkpointId;
+            lastCheckpointId = checkpointId;
+        } else {
+            throw new IllegalStateException(
+                    String.format(
+                            "Regressing checkpoint IDs. Previous checkpointId = %d, new checkpointId = %d",
+                            lastCheckpointId, checkpointId));
+        }
+    }
+
+    /**
+     * Closes the gateway. All events sent through this gateway are blocked until the gateway is
+     * re-opened. If the gateway is already closed, this does nothing.
+     *
+     * @return True if the gateway is closed, false if the checkpointId is incorrect.
+     */
+    boolean tryCloseGateway(long checkpointId) {
+        if (checkpointId == currentCheckpointId) {
+            isClosed = true;
+            return true;
+        }
+        return false;
+    }
+
+    boolean isClosed() {
+        return isClosed;
+    }
+
+    void openGatewayAndUnmarkCheckpoint(long expectedCheckpointId) {
+        checkRunsInMainThread();
+
+        if (expectedCheckpointId != currentCheckpointId) {
+            return;
+        }
+        openGatewayAndUnmarkCheckpoint();
+    }
+
+    /** Opens the gateway, releasing all buffered events. */
+    void openGatewayAndUnmarkCheckpoint() {
+        checkRunsInMainThread();
+
+        currentCheckpointId = NO_CHECKPOINT;
+        if (!isClosed) {
+            return;
+        }
+
+        for (BlockedEvent blockedEvent : blockedEvents) {
+            CompletableFuture<Acknowledge> result =
+                    innerGateway.sendEvent(blockedEvent.operatorEvent);
+            FutureUtils.forward(result, blockedEvent.future);
+        }
+        blockedEvents.clear();
+
+        isClosed = false;
+    }
+
+    @Override
+    public CompletableFuture<Acknowledge> sendEvent(OperatorEvent evt) {
+        if (isClosed) {
+            CompletableFuture<Acknowledge> sendResult = new CompletableFuture<>();
```

Comment:
Suppose the given operator event can not be serialized and the gateway is closed, the exception will be added to the future returned to the caller. 

This behavior is inconsistent with the case when the gateway is open. Also note that many callers currently do not check the future returned by this method. This means that an exception that can be caught previously might be missed now.

Would it be safer to keep the previous behavior by always throwing the deserialization exception directly to the caller?

Same for the exception that is thrown when the gateway is not read.

Reply:
You are right. I'll change the behavior here to always throw the exceptions if any.